File: runtime/ruby/lib/agentic_engine/llm.rb
# frozen_string_literal: true

require "openai"
require "json"
require "net/http"
require "uri"

module AgenticEngine
  # LLM interaction layer supporting OpenAI and Anthropic providers.
  # Uses the ruby-openai gem for OpenAI and raw Net::HTTP for Anthropic.
  module LLM
    class LLMError < StandardError; end

    ANTHROPIC_ENDPOINT = "https://api.anthropic.com/v1/messages"

    @log = Logger.create("llm")

    class << self
      # Call an LLM via the appropriate provider.
      #
      # @param model [String] e.g. "gpt-4.1" or "claude-sonnet-4-5-20241022"
      # @param system_prompt [String, nil] system message content
      # @param user_content [String] user message content
      # @param temperature [Float] sampling temperature (default 0.7)
      # @param schema_name [String, nil] if non-nil, load the named schema and
      #   request structured output; otherwise fall back to JSON mode.
      # @param base_url [String, nil] optional OpenAI-compatible base URL
      # @param api_key_env [String, nil] env var name holding the API key for base_url
      # @return [Hash] { output: <parsed JSON>, metrics: StepMetrics }
      def call_llm(model:, system_prompt: nil, user_content:, temperature: 0.7, schema_name: nil, base_url: nil, api_key_env: nil)
        if base_url
          @log.debug("call_llm", { provider: "openai-compatible", base_url: base_url, model: model, schema: schema_name, temperature: temperature })
          call_openai(
            model: model,
            system_prompt: system_prompt,
            user_content: user_content,
            temperature: temperature,
            schema_name: schema_name,
            base_url: base_url,
            api_key_env: api_key_env
          )
        elsif model&.start_with?("claude-")
          @log.debug("call_llm", { provider: "anthropic", model: model, schema: schema_name, temperature: temperature })
          call_anthropic(
            model: model,
            system_prompt: system_prompt,
            user_content: user_content,
            temperature: temperature,
            schema_name: schema_name
          )
        else
          @log.debug("call_llm", { provider: "openai", model: model, schema: schema_name, temperature: temperature })
          call_openai(
            model: model,
            system_prompt: system_prompt,
            user_content: user_content,
            temperature: temperature,
            schema_name: schema_name
          )
        end
      end

      private

      # ── OpenAI implementation ──

      def call_openai(model:, system_prompt:, user_content:, temperature:, schema_name:, base_url: nil, api_key_env: nil)
        messages = build_openai_messages(system_prompt: system_prompt, user_content: user_content)
        params = {
          model: model,
          messages: messages,
          temperature: temperature
        }

        if schema_name
          schema_def = Loader.load_schema(schema_name)
          params[:response_format] = {
            type: "json_schema",
            json_schema: {
              name: schema_def["name"],
              strict: schema_def.fetch("strict", true),
              schema: schema_def["schema"]
            }
          }
        else
          params[:response_format] = { type: "json_object" }
        end

        client = if base_url
                   api_key = api_key_env ? ENV[api_key_env] || "not-needed" : "not-needed"
                   OpenAI::Client.new(access_token: api_key, uri_base: base_url)
                 else
                   openai_client
                 end

        start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)
        response = client.chat(parameters: params)
        end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)

        choice = response.dig("choices", 0)
        raise LLMError, "No choices returned from OpenAI" unless choice

        finish_reason = choice["finish_reason"]
        if finish_reason == "refusal"
          refusal = choice.dig("message", "refusal")
          raise LLMError, "OpenAI refused request: #{refusal}"
        end

        raw_content = choice.dig("message", "content")
        raise LLMError, "Empty content from OpenAI" if raw_content.nil? || raw_content.strip.empty?

        parsed = JSON.parse(raw_content)

        usage = response["usage"] || {}
        metrics = StepMetrics.new(
          latency_ms:    (end_time - start_time).to_i,
          input_tokens:  usage["prompt_tokens"] || 0,
          output_tokens: usage["completion_tokens"] || 0
        )

        @log.info("OpenAI response", {
          model: model,
          latency_ms: metrics.latency_ms,
          input_tokens: metrics.input_tokens,
          output_tokens: metrics.output_tokens
        })

        { output: parsed, metrics: metrics }
      rescue JSON::ParserError => e
        raise LLMError, "Failed to parse OpenAI JSON response: #{e.message}"
      end

      # ── Anthropic implementation ──

      def call_anthropic(model:, system_prompt:, user_content:, temperature:, schema_name:)
        system = build_anthropic_system_prompt(system_prompt, schema_name)

        body = {
          model: model,
          max_tokens: 4096,
          system: system,
          messages: [{ role: "user", content: user_content }],
          temperature: temperature
        }

        start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)

        uri = URI.parse(ANTHROPIC_ENDPOINT)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        request = Net::HTTP::Post.new(uri.path)
        request["Content-Type"] = "application/json"
        request["x-api-key"] = anthropic_api_key
        request["anthropic-version"] = "2023-06-01"
        request.body = JSON.generate(body)

        response = http.request(request)
        end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC, :millisecond)

        unless response.is_a?(Net::HTTPSuccess)
          raise LLMError, "Anthropic API returned status #{response.code}: #{response.body}"
        end

        parsed_response = JSON.parse(response.body)

        # Extract text from content blocks
        text_block = (parsed_response["content"] || []).find { |b| b["type"] == "text" }
        raise LLMError, "No text content in Anthropic response" unless text_block

        raw_content = text_block["text"]
        raise LLMError, "Empty text content from Anthropic" if raw_content.nil? || raw_content.strip.empty?

        parsed = JSON.parse(raw_content)

        usage = parsed_response["usage"] || {}
        metrics = StepMetrics.new(
          latency_ms:    (end_time - start_time).to_i,
          input_tokens:  usage["input_tokens"] || 0,
          output_tokens: usage["output_tokens"] || 0
        )

        @log.info("Anthropic response", {
          model: model,
          latency_ms: metrics.latency_ms,
          input_tokens: metrics.input_tokens,
          output_tokens: metrics.output_tokens
        })

        { output: parsed, metrics: metrics }
      rescue JSON::ParserError => e
        raise LLMError, "Failed to parse Anthropic JSON response: #{e.message}"
      end

      # ── Helpers ──

      def openai_client
        @openai_client ||= OpenAI::Client.new(access_token: ENV.fetch("OPENAI_API_KEY") {
          raise LLMError, "OPENAI_API_KEY environment variable is not set"
        })
      end

      def anthropic_api_key
        @anthropic_api_key ||= ENV.fetch("ANTHROPIC_API_KEY") {
          raise LLMError, "ANTHROPIC_API_KEY environment variable is not set"
        }
      end

      def build_openai_messages(system_prompt:, user_content:)
        messages = []
        messages << { role: "system", content: system_prompt } if system_prompt && !system_prompt.empty?
        messages << { role: "user", content: user_content }
        messages
      end

      def build_anthropic_system_prompt(base, schema_name)
        base_prompt = base || ""

        if schema_name
          begin
            schema_def = Loader.load_schema(schema_name)
            schema_body = schema_def["schema"] || schema_def
            schema_json = JSON.pretty_generate(schema_body)
            return "#{base_prompt}\n\nYou must respond with a valid JSON object that conforms to the following JSON schema:\n#{schema_json}\n\nOutput only the JSON with no additional text, markdown, or code fences."
          rescue StandardError
            # Fall through to basic JSON instruction
          end
        end

        "#{base_prompt}\n\nYou must respond with a valid JSON object. Output only the JSON with no additional text, markdown, or code fences."
      end
    end
  end
end

File: runtime/python/agentic_engine/llm.py
"""LLM interaction layer supporting OpenAI and Anthropic providers.

Supports both structured-output mode (JSON Schema in system prompt for Anthropic,
``json_schema`` response format for OpenAI) and plain JSON-object mode, as well
as text and base64-image inputs.
"""

from __future__ import annotations

import json
import os
import time
from typing import Any

from anthropic import AsyncAnthropic
from openai import AsyncOpenAI

from .loader import load_schema
from .logger import create_logger, serialize_error
from .types import StepMetrics

log = create_logger("llm")

# Module-level clients — lazily initialised so the import itself never fails
# when API keys are not yet set.
_openai_client: AsyncOpenAI | None = None
_anthropic_client: AsyncAnthropic | None = None


def _get_openai_client() -> AsyncOpenAI:
    global _openai_client
    if _openai_client is None:
        _openai_client = AsyncOpenAI()
    return _openai_client


def _get_anthropic_client() -> AsyncAnthropic:
    global _anthropic_client
    if _anthropic_client is None:
        _anthropic_client = AsyncAnthropic()
    return _anthropic_client


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

async def call_llm(
    model: str,
    system_prompt: str,
    user_content: str | dict[str, Any],
    temperature: float = 0.0,
    schema_name: str | None = None,
    input_type: str = "text",
    image_detail: str = "auto",
    base_url: str = "",
    api_key_env: str = "",
) -> tuple[dict[str, Any], StepMetrics]:
    """Call an LLM and return ``(parsed_output, metrics)``.

    Parameters
    ----------
    model:
        Model identifier (e.g. ``gpt-4.1`` or ``claude-sonnet-4-5-20241022``).
    system_prompt:
        System-level instructions.
    user_content:
        The user message body.  For text input this is a plain string.  For
        image input this should be a base64-encoded data URI string.
    temperature:
        Sampling temperature.
    schema_name:
        If set, the response is requested as structured output conforming to
        the named JSON schema.  If ``None``, plain JSON mode is used.
    input_type:
        ``"text"`` (default) or ``"image"``.
    image_detail:
        Vision detail level (``"low"``, ``"high"``, ``"auto"``).  Only used
        when *input_type* is ``"image"``.
    base_url:
        Optional base URL for an OpenAI-compatible API endpoint.
    api_key_env:
        Environment variable name to read the API key from when *base_url* is
        set.  If empty (and base_url is set), ``"not-needed"`` is used.
    """
    if base_url:
        api_key = os.environ.get(api_key_env, "not-needed") if api_key_env else "not-needed"
        client = AsyncOpenAI(base_url=base_url, api_key=api_key)
        log.info(
            "Calling LLM",
            provider="openai-compatible",
            base_url=base_url,
            model=model,
            schema_name=schema_name,
            temperature=temperature,
            input_type=input_type,
        )
        return await _call_openai(
            model=model,
            system_prompt=system_prompt,
            user_content=user_content,
            temperature=temperature,
            schema_name=schema_name,
            input_type=input_type,
            image_detail=image_detail,
            client=client,
        )
    elif model.startswith("claude-"):
        log.info(
            "Calling LLM",
            provider="anthropic",
            model=model,
            schema_name=schema_name,
            temperature=temperature,
            input_type=input_type,
        )
        return await _call_anthropic(
            model=model,
            system_prompt=system_prompt,
            user_content=user_content,
            temperature=temperature,
            schema_name=schema_name,
            input_type=input_type,
            image_detail=image_detail,
        )
    else:
        log.info(
            "Calling LLM",
            provider="openai",
            model=model,
            schema_name=schema_name,
            temperature=temperature,
            input_type=input_type,
        )
        return await _call_openai(
            model=model,
            system_prompt=system_prompt,
            user_content=user_content,
            temperature=temperature,
            schema_name=schema_name,
            input_type=input_type,
            image_detail=image_detail,
        )


# ---------------------------------------------------------------------------
# OpenAI implementation
# ---------------------------------------------------------------------------

async def _call_openai(
    model: str,
    system_prompt: str,
    user_content: str | dict[str, Any],
    temperature: float,
    schema_name: str | None,
    input_type: str,
    image_detail: str,
    client: AsyncOpenAI | None = None,
) -> tuple[dict[str, Any], StepMetrics]:
    client = client or _get_openai_client()

    # -- Build messages -------------------------------------------------------
    messages: list[dict[str, Any]] = []

    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})

    if input_type == "image" and isinstance(user_content, str):
        image_url = user_content
        if not user_content.startswith("data:"):
            image_url = f"data:image/png;base64,{user_content}"
        messages.append({
            "role": "user",
            "content": [
                {
                    "type": "image_url",
                    "image_url": {"url": image_url, "detail": image_detail},
                }
            ],
        })
    else:
        text = user_content if isinstance(user_content, str) else json.dumps(user_content, default=str)
        messages.append({"role": "user", "content": text})

    # -- Build response_format ------------------------------------------------
    response_format: dict[str, Any]
    if schema_name is not None:
        schema_obj = load_schema(schema_name)
        response_format = {
            "type": "json_schema",
            "json_schema": {
                "name": schema_obj.get("name", schema_name),
                "schema": schema_obj.get("schema", schema_obj),
                "strict": schema_obj.get("strict", True),
            },
        }
    else:
        response_format = {"type": "json_object"}

    # -- Call the API ---------------------------------------------------------
    start = time.monotonic()
    try:
        response = await client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=temperature,
            response_format=response_format,
        )
    except Exception as exc:
        ei = serialize_error(exc)
        log.error("OpenAI call failed", error=ei["message"], error_name=ei["name"])
        raise

    elapsed_ms = (time.monotonic() - start) * 1000

    # -- Parse response -------------------------------------------------------
    choice = response.choices[0]
    raw_text = choice.message.content or "{}"

    try:
        output = json.loads(raw_text)
    except json.JSONDecodeError:
        log.warn("OpenAI returned non-JSON content, wrapping as text", raw=raw_text[:200])
        output = {"raw": raw_text}

    usage = response.usage
    metrics = StepMetrics(
        duration_ms=round(elapsed_ms, 2),
        prompt_tokens=usage.prompt_tokens if usage else 0,
        completion_tokens=usage.completion_tokens if usage else 0,
        total_tokens=usage.total_tokens if usage else 0,
    )

    log.info(
        "OpenAI call completed",
        duration_ms=metrics.duration_ms,
        total_tokens=metrics.total_tokens,
    )

    return output, metrics


# ---------------------------------------------------------------------------
# Anthropic implementation
# ---------------------------------------------------------------------------

def _build_anthropic_system_prompt(
    base: str, schema_name: str | None
) -> str:
    """Augment the system prompt with JSON output instructions.

    When *schema_name* is provided the corresponding JSON schema is embedded
    so the model returns structured output.
    """
    if schema_name is not None:
        try:
            schema_obj = load_schema(schema_name)
            schema_body = schema_obj.get("schema", schema_obj)
            schema_json = json.dumps(schema_body, indent=2)
            return (
                f"{base}\n\nYou must respond with a valid JSON object that "
                f"conforms to the following JSON schema:\n{schema_json}\n\n"
                "Output only the JSON with no additional text, markdown, or code fences."
            )
        except Exception:
            pass  # Fall through to basic JSON instruction

    return (
        f"{base}\n\nYou must respond with a valid JSON object. "
        "Output only the JSON with no additional text, markdown, or code fences."
    )


def _build_anthropic_messages(
    user_content: str | dict[str, Any],
    input_type: str,
    image_detail: str,
) -> list[dict[str, Any]]:
    """Build Anthropic-format messages from user content."""
    if input_type == "image" and isinstance(user_content, str):
        image_data = user_content
        media_type = "image/png"
        # Parse data URI if present
        if image_data.startswith("data:"):
            import re
            match = re.match(r"^data:(image/\w+);base64,(.+)$", image_data)
            if match:
                media_type = match.group(1)
                image_data = match.group(2)
        return [{
            "role": "user",
            "content": [
                {
                    "type": "image",
                    "source": {
                        "type": "base64",
                        "media_type": media_type,
                        "data": image_data,
                    },
                }
            ],
        }]

    text = user_content if isinstance(user_content, str) else json.dumps(user_content, default=str)
    return [{"role": "user", "content": text}]


async def _call_anthropic(
    model: str,
    system_prompt: str,
    user_content: str | dict[str, Any],
    temperature: float,
    schema_name: str | None,
    input_type: str,
    image_detail: str,
) -> tuple[dict[str, Any], StepMetrics]:
    client = _get_anthropic_client()

    system = _build_anthropic_system_prompt(system_prompt, schema_name)
    messages = _build_anthropic_messages(user_content, input_type, image_detail)

    start = time.monotonic()
    try:
        response = await client.messages.create(
            model=model,
            max_tokens=4096,
            system=system,
            messages=messages,
            temperature=temperature,
        )
    except Exception as exc:
        ei = serialize_error(exc)
        log.error("Anthropic call failed", error=ei["message"], error_name=ei["name"])
        raise

    elapsed_ms = (time.monotonic() - start) * 1000

    # Extract text from content blocks
    raw_text = ""
    for block in response.content:
        if block.type == "text":
            raw_text = block.text
            break

    if not raw_text:
        raw_text = "{}"

    # Strip markdown code fences if present (Claude often wraps JSON in ```json ... ```)
    stripped = raw_text.strip()
    if stripped.startswith("```"):
        lines = stripped.split("\n")
        # Remove first line (```json or ```) and last line (```)
        lines = lines[1:]
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]
        stripped = "\n".join(lines).strip()

    try:
        output = json.loads(stripped)
    except json.JSONDecodeError:
        log.warn("Anthropic returned non-JSON content, wrapping as text", raw=raw_text[:200])
        output = {"raw": raw_text}

    usage = response.usage
    metrics = StepMetrics(
        duration_ms=round(elapsed_ms, 2),
        prompt_tokens=usage.input_tokens if usage else 0,
        completion_tokens=usage.output_tokens if usage else 0,
        total_tokens=(usage.input_tokens + usage.output_tokens) if usage else 0,
    )

    log.info(
        "Anthropic call completed",
        duration_ms=metrics.duration_ms,
        total_tokens=metrics.total_tokens,
    )

    return output, metrics
